packages/phishing-controller/src/utils.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  Hotlist,
+  PhishingListState,
+  PhishingStalelist,
+} from './PhishingController';
+
+/**
+ * Composes an object mapping from domain to it's inclusion in list state. This is used to initialize the
+ * list state's domains in a map such that removals or further additions can be applied.
+ *
+ * @param list - a list of domains
+ * @returns an object mapping between domain and whether or not that domain is to be included in new list state.
+ */
+export const buildListMap = (list: string[]): Record<string, true> =>
+  list.reduce((acc, cur) => {
+    acc[cur] = true;
+    return acc;
+  }, {} as Record<string, true>);
+
+/**
+ * Converts the listState into an object of objects defined by list type and specific domain to determine
+ * whether a domain should be included in listState or not.
+ *
+ * @param map - an object mapping between domain and whether or not that domain is to be included in new list state.
+ * @returns a list of domains that are included in new liststate
+ */
+export const convertMapToList = (map: Record<string, boolean>): string[] =>
+  Object.entries(map).reduce<string[]>((acc, [domain, isIncluded]) => {
+    if (isIncluded) {
+      acc.push(domain);
+    }
+    return acc;
+  }, []);
+
+/**
+ * Fetches current epoch time in seconds.
+ *
+ * @returns the Date.now() time in seconds instead of miliseconds. backend files rely on timestamps in seconds since epoch.
+ */
+export const fetchTimeNow = (): number => Math.round(Date.now() / 1000);
+
+/**
+ * Converts the listState into an object of objects defined by list type and specific domain to determine
+ * whether a domain should be included in listState or not.
+ *
+ * @param listState - the stalelist or the existing liststate that diffs will be applied to.
+ * @returns the domain inclusion map on a per-list-type basis
+ */
+export const convertListStateToMap = (listState: PhishingListState) => ({
+  allowlist: buildListMap(listState.allowlist),
+  fuzzylist: buildListMap(listState.fuzzylist),
+  blocklist: buildListMap(listState.blocklist),
+});
+
+/**
+ * Determines which diffs are applicable to the listState, then applies those diffs.
+ *
+ * @param listState - the stalelist or the existing liststate that diffs will be applied to.
+ * @param hotlistDiffs - the diffs to apply to the listState if valid.
+ * @returns the new list state
+ */
+export const applyDiffs = (
+  listState: PhishingListState | PhishingStalelist,
+  hotlistDiffs: Hotlist,
+): PhishingListState => {
+  const diffsToApply = hotlistDiffs.filter(
+    ({ timestamp }) => timestamp > listState.lastUpdated,
+  );
+
+  const map = convertListStateToMap(listState);
```

Comment:
Hmm. Thinking of the per issues we just had recently, I am a bit concerned about this data manipulation step.

Converting to an object does intuitively seem like it would be faster than using `.includes`, so that's good. But I think something like this would work as well using Sets:


```
export const applyDiffs = (
  listState: PhishingListState | PhishingStalelist,
  hotlistDiffs: Hotlist,
): PhishingListState => {
  const diffsToApply = hotlistDiffs.filter(
    ({ timestamp }) => timestamp > listState.lastUpdated,
  );

  const urlsToRemove = {
    allowlist: new Set(),
    blocklist: new Set(),
    fuzzylist: new Set(),
  };
  const urlsToAdd = {
    allowlist: new Set(),
    blocklist: new Set(),
    fuzzylist: new Set(),
  };
  for (const { isRemoval, targetList, url } of diffsToApply) {
    if (isRemoval) {
      urlsToRemove[targetList].add(url);
    } else {
      urlsToAdd[targetList].add(url);
    }
  }

  const listSets = {
    allowlist: new Set(),
    blocklist: new Set(),
    fuzzylist: new Set(),
  };
  for (const listName of ['allowlist', 'blocklist', 'fuzzylist']) {
    for (const entryToDelete of urlsToRemove[listName]) {
      listSets[listName].delete(entryToDelete);
    }
    for (const entryToAdd of urlsToAdd[listName]) {
      listSets[listName].add(entryToAdd);
    }
  }

  const now = fetchTimeNow();

  return {
    allowlist: Array.from(listSets.allowlist),
    blocklist: Array.from(listSets.blocklist),
    fuzzylist: Array.from(listSets.fuzzylist),
    version: listState.version,
    tolerance: listState.tolerance,
    name: 'MetaMask',
    lastUpdated: now,
  };
};
```

Just a quick sketch, the types don't even work yet. I think it doesn't like the string references to the list name.

Speaking which, can we use an enum for those list names? That light help TypeScript out a bit and get rid of the need for a few casts I've seen here and there.

Reply:
Benchmark seems to indicate using a set is ~twice as fast: https://perf.link/#eyJpZCI6ImJ0ZG9tM2QwbDhpIiwidGl0bGUiOiJBcnJheS5pbmNsdWRlcyB2cy4gU2V0LmhhcyIsImJlZm9yZSI6ImNvbnN0IGxpc3RTaXplID0gMjAwMDA7XG5jb25zdCBsaXN0U3RhdGUgPSB7XG4gIGFsbG93bGlzdDogWy4uLm5ldyBBcnJheShsaXN0U2l6ZSldLm1hcCgoKSA9PiBgJHtNYXRoLnJhbmRvbSgpfS5jb21gKSxcbiAgYmxvY2tsaXN0OiBbLi4ubmV3IEFycmF5KGxpc3RTaXplKV0ubWFwKCgpID0%2BIGAke01hdGgucmFuZG9tKCl9LmNvbWApLFxuICBmdXp6eWxpc3Q6IFsuLi5uZXcgQXJyYXkobGlzdFNpemUpXS5tYXAoKCkgPT4gYCR7TWF0aC5yYW5kb20oKX0uY29tYCksXG59XG5jb25zdCBkaWZmc1RvQXBwbHkgPSBbLi4ubmV3IEFycmF5KGxpc3RTaXplLyAxMDApXS5tYXAoKCkgPT4ge1xuICBjb25zdCByYW5kb21OdW1iZXIgPSBNYXRoLnJhbmRvbSgpO1xuICByZXR1cm4ge1xuICAgIGlzUmVtb3ZhbDogcmFuZG9tTnVtYmVyID4gMC41LFxuICAgIHVybDogYCR7cmFuZG9tTnVtYmVyfS5jb21gLFxuICAgIHRhcmdldExpc3Q6IHJhbmRvbU51bWJlciA%2BIDAuMzMzMzMgP1xuICAgICAgcmFuZG9tTnVtYmVyID4gMC42NjY2NiA%2FXG4gICAgICAgICdhbGxvd2xpc3QnIDpcbiAgICAgICAgJ2Jsb2NrbGlzdCcgOlxuICAgICAgJ2Z1enp5bGlzdCdcbiAgfVxufSkiLCJ0ZXN0cyI6W3sibmFtZSI6IkFycmF5LmluY2x1ZGVzIiwiY29kZSI6ImNvbnN0IGJ1aWxkTGlzdE1hcCA9IChsaXN0KSA9PlxuICBsaXN0LnJlZHVjZSgoYWNjLCBjdXIpID0%2BIHtcbiAgICBhY2NbY3VyXSA9IHRydWU7XG4gICAgcmV0dXJuIGFjYztcbiAgfSwge30pO1xuICBcbmNvbnN0IGNvbnZlcnRNYXBUb0xpc3QgPSAobWFwKSA9PlxuICBPYmplY3QuZW50cmllcyhtYXApLnJlZHVjZSgoYWNjLCBbZG9tYWluLCBpc0luY2x1ZGVkXSkgPT4ge1xuICAgIGlmIChpc0luY2x1ZGVkKSB7XG4gICAgICBhY2MucHVzaChkb21haW4pO1xuICAgIH1cbiAgICByZXR1cm4gYWNjO1xuICB9LCBbXSk7XG5cbmNvbnN0IGNvbnZlcnRMaXN0U3RhdGVUb01hcCA9IChsaXN0U3RhdGUpID0%2BICh7XG4gIGFsbG93bGlzdDogYnVpbGRMaXN0TWFwKGxpc3RTdGF0ZS5hbGxvd2xpc3QpLFxuICBmdXp6eWxpc3Q6IGJ1aWxkTGlzdE1hcChsaXN0U3RhdGUuZnV6enlsaXN0KSxcbiAgYmxvY2tsaXN0OiBidWlsZExpc3RNYXAobGlzdFN0YXRlLmJsb2NrbGlzdCksXG59KTtcblxuY29uc3QgbWFwID0gY29udmVydExpc3RTdGF0ZVRvTWFwKGxpc3RTdGF0ZSk7XG5jb25zdCBvdXRwdXRNYXAgPSBkaWZmc1RvQXBwbHkucmVkdWNlKChhY2MsIGN1cnJEaWZmKSA9PiB7XG4gIGFjY1tjdXJyRGlmZi50YXJnZXRMaXN0XVtjdXJyRGlmZi51cmxdID0gIWN1cnJEaWZmLmlzUmVtb3ZhbDtcbiAgcmV0dXJuIGFjYztcbn0sIG1hcCk7XG5cbmNvbnN0IHJlc3VsdCA9IHtcbiAgYWxsb3dsaXN0OiBjb252ZXJ0TWFwVG9MaXN0KG91dHB1dE1hcC5hbGxvd2xpc3QpLFxuICBibG9ja2xpc3Q6IGNvbnZlcnRNYXBUb0xpc3Qob3V0cHV0TWFwLmJsb2NrbGlzdCksXG4gIGZ1enp5bGlzdDogY29udmVydE1hcFRvTGlzdChvdXRwdXRNYXAuZnV6enlsaXN0KSxcbn07IiwicnVucyI6WzQ3LDQyLDQyLDQyLDQyLDUyLDQ3LDQyLDQ3LDQyLDQ3LDQ3LDQ3LDQ3LDUyLDQ3LDQ3LDQ3LDQ3LDQ3LDQ3LDQ3LDQ3LDQ3LDQ3LDQyLDQyLDQ3LDQ3LDQyLDQ3LDQyLDQyLDQyLDQyLDQ3LDM4LDQ3LDQyLDQ3LDQ3LDQ3LDQyLDQ3LDQ3LDQ3LDQyLDQyLDQ3LDQ3LDQyLDQyLDQ3LDQyLDQ3LDQyLDQyLDQ3LDM4LDQyLDQyLDQyLDQyLDQyLDQ3LDQyLDQyLDQyLDQyLDQ3LDQyLDQyLDQ3LDQyLDQyLDQyLDQyLDQyLDQyLDQ3LDQ3LDQ3LDQyLDQyLDQ3LDQ3LDQ3LDQ3LDQ3LDQyLDQyLDQ3LDQ3LDQ3LDQ3LDQ3LDQ3LDQyLDQyLDQ3XSwib3BzIjo0NH0seyJuYW1lIjoiU2V0LmhhcyIsImNvZGUiOiJjb25zdCB1cmxzVG9SZW1vdmUgPSB7XG4gIGFsbG93bGlzdDogbmV3IFNldCgpLFxuICBibG9ja2xpc3Q6IG5ldyBTZXQoKSxcbiAgZnV6enlsaXN0OiBuZXcgU2V0KCksXG59O1xuY29uc3QgdXJsc1RvQWRkID0ge1xuICBhbGxvd2xpc3Q6IG5ldyBTZXQoKSxcbiAgYmxvY2tsaXN0OiBuZXcgU2V0KCksXG4gIGZ1enp5bGlzdDogbmV3IFNldCgpLFxufTtcbmZvciAoY29uc3QgeyBpc1JlbW92YWwsIHRhcmdldExpc3QsIHVybCB9IG9mIGRpZmZzVG9BcHBseSkge1xuICBpZiAoaXNSZW1vdmFsKSB7XG4gICAgdXJsc1RvUmVtb3ZlW3RhcmdldExpc3RdLmFkZCh1cmwpO1xuICB9IGVsc2Uge1xuICAgIHVybHNUb0FkZFt0YXJnZXRMaXN0XS5hZGQodXJsKTtcbiAgfVxufVxuXG5jb25zdCBsaXN0U2V0cyA9IHtcbiAgYWxsb3dsaXN0OiBuZXcgU2V0KGxpc3RTdGF0ZS5hbGxvd2xpc3QpLFxuICBibG9ja2xpc3Q6IG5ldyBTZXQobGlzdFN0YXRlLmJsb2NrbGlzdCksXG4gIGZ1enp5bGlzdDogbmV3IFNldChsaXN0U3RhdGUuZnV6enlsaXN0KSxcbn07XG5mb3IgKGNvbnN0IGxpc3ROYW1lIG9mIFsnYWxsb3dsaXN0JywgJ2Jsb2NrbGlzdCcsICdmdXp6eWxpc3QnXSkge1xuICBmb3IgKGNvbnN0IGVudHJ5VG9EZWxldGUgb2YgdXJsc1RvUmVtb3ZlW2xpc3ROYW1lXSkge1xuICAgIGxpc3RTZXRzW2xpc3ROYW1lXS5kZWxldGUoZW50cnlUb0RlbGV0ZSk7XG4gIH1cbiAgZm9yIChjb25zdCBlbnRyeVRvQWRkIG9mIHVybHNUb0FkZFtsaXN0TmFtZV0pIHtcbiAgICBsaXN0U2V0c1tsaXN0TmFtZV0uYWRkKGVudHJ5VG9BZGQpO1xuICB9XG59XG5cbmNvbnN0IHJlc3VsdCA9IHtcbiAgYWxsb3dsaXN0OiBBcnJheS5mcm9tKGxpc3RTZXRzLmFsbG93bGlzdCksXG4gIGJsb2NrbGlzdDogQXJyYXkuZnJvbShsaXN0U2V0cy5ibG9ja2xpc3QpLFxuICBmdXp6eWxpc3Q6IEFycmF5LmZyb20obGlzdFNldHMuZnV6enlsaXN0KSxcbn07IiwicnVucyI6Wzg1LDgwLDgwLDc2LDc2LDg1LDkwLDgwLDg1LDc2LDgwLDgwLDg1LDgwLDg1LDg1LDgwLDg1LDg1LDgwLDg1LDgwLDg1LDg1LDcxLDc2LDc2LDg1LDgwLDc2LDc2LDc2LDgwLDc2LDc2LDc2LDY2LDgwLDcxLDkwLDgwLDgwLDgwLDgwLDg1LDgwLDgwLDcxLDg1LDc2LDcxLDc2LDgwLDc2LDgwLDgwLDc2LDgwLDcxLDcxLDcxLDc2LDcxLDc2LDgwLDcxLDcxLDc2LDc2LDgwLDc2LDc2LDgwLDc2LDc2LDcxLDcxLDc2LDc2LDg1LDg1LDc2LDgwLDgwLDgwLDgwLDgwLDc2LDg1LDc2LDgwLDg1LDgwLDgwLDg1LDgwLDgwLDc2LDc2LDg1XSwib3BzIjo3OH1dLCJ1cGRhdGVkIjoiMjAyMy0wMi0wNlQxODoxMToyMC42NThaIn0%3D

Reply:
Here is a much more concise version of what I suggested above:

```typescript
export const applyDiffs = (
  listState: PhishingListState | PhishingStalelist,
  hotlistDiffs: Hotlist,
): PhishingListState => {
  const diffsToApply = hotlistDiffs.filter(
    ({ timestamp }) => timestamp > listState.lastUpdated,
  );

  const listSets = {
    allowlist: new Set(listState.allowlist),
    blocklist: new Set(listState.blocklist),
    fuzzylist: new Set(listState.fuzzylist),
  };
  
  for (const { isRemoval, targetList, url } of diffsToApply) {
    if (isRemoval) {
      listSets[targetList].delete(url);
    } else {
      listSets[targetList].add(url);
    }
  }

  const now = fetchTimeNow();

  return {
    allowlist: Array.from(listSets.allowlist),
    blocklist: Array.from(listSets.blocklist),
    fuzzylist: Array.from(listSets.fuzzylist),
    version: listState.version,
    tolerance: listState.tolerance,
    name: 'MetaMask',
    lastUpdated: now,
  };
};
```